File: validate_binary_search/validate_binary_search.rb
require 'rspec'

class TreeNode
  attr_accessor :val, :left, :right

  def initialize(val = 0, left = nil, right = nil)
    @val = val
    @left = left
    @right = right
  end
end

class Solution
  def is_valid_bst(root)
    valid_bst?(root, -Float::INFINITY, Float::INFINITY)
  end

  private

  def valid_bst?(node, min, max)
    return true if node.nil?

    return false if node.val <= min || node.val >= max

    valid_bst?(node.left, min, node.val) && valid_bst?(node.right, node.val, max)
  end
end

RSpec.describe Solution do
  let(:solution) { Solution.new }

  describe '#is_valid_bst' do
    it 'returns true for a valid BST' do
      # Creating a valid BST: 
      #       2
      #      / \
      #     1   3
      root = TreeNode.new(2)
      root.left = TreeNode.new(1)
      root.right = TreeNode.new(3)

      expect(solution.is_valid_bst(root)).to be true
    end

    it 'returns false for an invalid BST' do
      # Creating an invalid BST:
      #       5
      #      / \
      #     1   4
      #        / \
      #       3   6
      root = TreeNode.new(5)
      root.left = TreeNode.new(1)
      root.right = TreeNode.new(4)
      root.right.left = TreeNode.new(3)
      root.right.right = TreeNode.new(6)

      expect(solution.is_valid_bst(root)).to be false
    end

    it 'returns true for an empty tree' do
      root = nil
      expect(solution.is_valid_bst(root)).to be true
    end
    it 'returns false when left child is greater than parent' do
      # Creating an invalid BST:
      #       2
      #      / \
      #     3   4
      root = TreeNode.new(2)
      root.left = TreeNode.new(3)
      root.right = TreeNode.new(4)

      expect(solution.is_valid_bst(root)).to be false
    end


  end
end

File: validate_binary_search/validate_binary_search.exs
defmodule Solution do
  def is_valid_bst(root) do
    valid_bst?(root, -:infinity, :infinity)
  end

  defp valid_bst?(nil, _min, _max), do: true

  defp valid_bst?(%TreeNode{val: val, left: left, right: right}, min, max) do
    if val <= min or val >= max do
      false
    else
      valid_bst?(left, min, val) and valid_bst?(right, val, max)
    end
  end
end

defmodule TreeNode do
  defstruct [:val, :left, :right]

  def new(val, left \\ nil, right \\ nil) do
    %TreeNode{val: val, left: left, right: right}
  end
end

defmodule SolutionTest do
  use ExUnit.Case
  alias Solution
  alias TreeNode

  test "valid BST" do
    root = %TreeNode{val: 2, left: %TreeNode{val: 1}, right: %TreeNode{val: 3}}
    assert Solution.is_valid_bst(root) == true
  end
  
  test "invalid BST" do
    root = %TreeNode{
      val: 5,
      left: %TreeNode{val: 1},
      right: %TreeNode{
        val: 4,
        left: %TreeNode{val: 3},
        right: %TreeNode{val: 6}
      }
    }
    assert Solution.is_valid_bst(root) == false
  end

  test "empty tree" do
    root = nil
    assert Solution.is_valid_bst(root) == true
  end

  test "left child greater than parent" do
    root = %TreeNode{val: 2, left: %TreeNode{val: 3}, right: %TreeNode{val: 4}}
    assert Solution.is_valid_bst(root) == false
  end

  test "right child smaller than parent" do
    root = %TreeNode{val: 2, left: %TreeNode{val: 1}, right: %TreeNode{val: 0}}
    assert Solution.is_valid_bst(root) == false
  end

end
